### PR TITLE
feat: decompose resources

### PR DESCRIPTION
**Pull request by felangel**
## Description

- [X] add `Artifact`, `Channel`, and `Release` objects to `package:code_push_protocol`
- [X] expose `createChannel`, `createRelease`, `createPatch`, `createArtifact`, 'getChannels', and `getReleases` from `package:code_push_client`
- [X] document intended usage in `package:shorebird_cli`
- [x] update `package:shorebird_cli` to align with proposed changes

## Type of Change

- [X] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/shorebird_code_push_protocol/lib/src/models/channel.dart [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import 'package:json_annotation/json_annotation.dart';
+
+part 'channel.g.dart';
+
+/// {@template channel}
+/// A tag used to manage the subset of applications that receive a patch.
```
Comment (eseidel) [inline packages/shorebird_code_push_protocol/lib/src/models/channel.dart]: subset of devices? installs?

**Review by eseidel — APPROVED** [review]

File: packages/shorebird_code_push_protocol/lib/src/models/channel.dart (hunk 1 above)
Comment (felangel) [reply packages/shorebird_code_push_protocol/lib/src/models/channel.dart]: yeah good call I just took what's in the glossary but will adjust both (I think subset of installs makes sense).

File: packages/shorebird_cli/lib/src/commands/publish_command.dart [hunk 2]
```diff
@@ -39,55 +55,197 @@ class PublishCommand extends ShorebirdCommand with ShorebirdConfigMixin {
...
+        );
+        createReleaseProgress.complete();
+      } catch (error) {
+        createReleaseProgress.fail('$error');
+        return ExitCode.software.code;
+      }
+    }
+
+    late final Patch patch;
+    final createPatchProgress = logger.progress('Creating patch');
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/commands/publish_command.dart:185]: I've found most times that I have a late variable the block setting the late probably should be a function instead.

In this case, it seems like you want a function helper to deal with this "finally" behavior of managing the progress indicator?  e.g.

final Patch patch = logger.indeterminant("Creating patch", () { await codePushClient.createPatch(..));

File: packages/shorebird_cli/lib/src/shorebird_config_mixin.dart [hunk 3]
```diff
@@ -6,6 +6,7 @@ import 'package:pubspec_parse/pubspec_parse.dart';
 import 'package:shorebird_cli/src/command.dart';
 import 'package:shorebird_cli/src/config/config.dart';
 import 'package:yaml/yaml.dart';
+import 'package:yaml_edit/yaml_edit.dart';
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/shorebird_config_mixin.dart:9]: I've used this package before, and I recall having a tickle in my brain about something being wrong with it. e.g. not maintaining comments or order or something. 🤷  Maybe whatever was wrong then is fixed now. :)

File: packages/shorebird_cli/lib/src/shorebird_config_mixin.dart [hunk 4]
```diff
@@ -48,4 +49,48 @@ mixin ShorebirdConfigMixin on ShorebirdCommand {
...
+  ShorebirdYaml addShorebirdYamlToProject(String appId) {
+    File(
+      p.join(Directory.current.path, 'shorebird.yaml'),
+    ).writeAsStringSync('''
+# This file is used to configure the Shorebird CLI.
+# Learn more at https://shorebird.dev
+
+# This is the unique identifier assigned to your app.
+# It is used by your app to request the correct patches from the Shorebird servers.
+app_id: $appId
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/shorebird_config_mixin.dart:62]: Eventually we will want to add commented out versions of the other parameters this supports to make this file self-documenting. :)

File: packages/shorebird_code_push_client/lib/src/code_push_client.dart [hunk 5]
```diff
@@ -46,6 +46,35 @@ class CodePushClient {
...
+    required String hash,
+  }) async {
+    final request = http.MultipartRequest(
+      'POST',
+      Uri.parse('$hostedUri/api/v1/artifacts'),
+    );
+    final file = await http.MultipartFile.fromPath('file', artifactPath);
+    request.files.add(file);
+    request.fields.addAll({
+      'patch_id': '$patchId',
```
Comment (eseidel) [inline packages/shorebird_code_push_client/lib/src/code_push_client.dart:64]: patchid is different from patch_number correct?

File: packages/shorebird_code_push_client/lib/src/code_push_client.dart [hunk 6]
```diff
@@ -62,34 +91,61 @@ class CodePushClient {
...
+    final response = await _httpClient.post(
+      Uri.parse('$hostedUri/api/v1/channels'),
+      headers: _apiKeyHeader,
+      body: json.encode({'app_id': appId, 'channel': channel}),
+    );
+
+    if (response.statusCode != HttpStatus.ok) {
+      throw _parseErrorResponse(response.body);
+    }
+    final body = json.decode(response.body) as Map<String, dynamic>;
```
Comment (eseidel) [inline packages/shorebird_code_push_client/lib/src/code_push_client.dart:108]: All this should be done for you. :(

File: packages/shorebird_code_push_client/lib/src/code_push_client.dart [hunk 7]
```diff
@@ -138,6 +194,60 @@ class CodePushClient {
         .toList();
   }
 
+  /// List all channels for the provided [appId].
+  Future<List<Channel>> getChannels({required String appId}) async {
+    final response = await _httpClient.get(
+      Uri.parse('$hostedUri/api/v1/channels').replace(
```
Comment (eseidel) [inline packages/shorebird_code_push_client/lib/src/code_push_client.dart:200]: The version probably doesn't belong in every single caller?  (A later problem to fix.)

**Review by eseidel — APPROVED** [review]

File: packages/shorebird_cli/lib/src/commands/publish_command.dart (hunk 2 above)
Comment (felangel) [reply packages/shorebird_cli/lib/src/commands/publish_command.dart:185]: Yeah that could work I'll mess around with that API and potentially refactor in a subsequent PR 👍

File: packages/shorebird_cli/lib/src/shorebird_config_mixin.dart (hunk 4 above)
Comment (felangel) [reply packages/shorebird_cli/lib/src/shorebird_config_mixin.dart:62]: Yup agreed

File: packages/shorebird_code_push_client/lib/src/code_push_client.dart (hunk 5 above)
Comment (felangel) [reply packages/shorebird_code_push_client/lib/src/code_push_client.dart:64]: Yup, patch_number is not unique and is relative to the release

File: packages/shorebird_code_push_client/lib/src/code_push_client.dart (hunk 6 above)
Comment (felangel) [reply packages/shorebird_code_push_client/lib/src/code_push_client.dart:108]: yeah agreed I had a p.o.c that does that for Dart Frog

File: packages/shorebird_code_push_client/lib/src/code_push_client.dart (hunk 7 above)
Comment (felangel) [reply packages/shorebird_code_push_client/lib/src/code_push_client.dart:200]: yeah agreed we can extract that in a subsequent PR 👍

**Pull request merged**